### PR TITLE
Implement SVD for Mat3x3

### PR DESCRIPTION
**Pull request by prikshitsingh24**
fixes: #220 
This is a port of : https://github.com/wi-re/tbtSVD/blob/master/source/SVD.h
It calculates V matrix correctly but the remaining S and U matrix are still not accurate.
Correct answer:
chrome_d60ehatq9n
Algorithm answer:
`
SVDSet { U: Mat3 { x_axis: Vec3(1.7881393e-7, 0.0, 0.9999999), y_axis: Vec3(0.0, -1.0, 0.0), z_axis: Vec3(0.9999999, 0.0, -1.7881393e-7) }, S: Mat3 { x_axis: Vec3(2.9999995, -0.0, 1.7881393e-7), y_axis: Vec3(0.0, 2.0, -0.0), z_axis: Vec3(-5.364418e-7, 0.0, 0.9999999) }, V: Mat3 { x_axis: Vec3(1.0, 0.0, 0.0), y_axis: Vec3(0.0, 1.0, -0.0), z_axis: Vec3(-0.0, 0.0, 1.0) } }
`

### CODE REVIEW

File: crates/kornia-linalg/src/linalg.rs [hunk 1]
```diff
@@ -0,0 +1,509 @@
...
+
+    Q.z_axis.x = 2.0 * g2.ch * g2.sh;
+    Q.z_axis.y = -2.0 * g3.ch * sh22 * g3.sh;
+    Q.z_axis.z = sh22 * sh32;
+
+    QR { Q, R }
+}
+
+/// Wrapping function used to contain all of the required sub calls
+pub fn svd(A: Mat3) -> SVDSet {
```
Comment (edgarriba) [inline crates/kornia-linalg/src/linalg.rs:451]: maybe call it `svd3` ?

File: crates/kornia-linalg/src/linalg.rs [hunk 2]
```diff
@@ -0,0 +1,509 @@
...
+
+    let w = rsqrt(g.ch * g.ch + g.sh * g.sh);
+    g.ch *= w;
+    g.sh *= w;
+
+    return g;
+}
+
+/// Implements a QR decomposition of a Matrix
+pub fn qr_decomposition(B: &mut Mat3) -> QR {
```
Comment (edgarriba) [inline crates/kornia-linalg/src/linalg.rs:377]:
```suggestion
fn qr_decomposition(B: &mut Mat3) -> QR {
```
make all the "utils" as private functions

**Review by edgarriba — COMMENTED** [review]
I'll review the math soon -- but good start !

File: crates/kornia-linalg/src/linalg.rs [hunk 3]
```diff
@@ -0,0 +1,509 @@
...
+}
+
+#[cfg(test)]
+mod tests {
+    use glam::{Mat3, Vec3};
+
+    use super::*;
+
+    #[test]
+    fn test_svd() {
```
Comment (edgarriba) [inline crates/kornia-linalg/src/linalg.rs:481]: can you add a numerical test ? to check that we do not have any regression ? You could als try to reconstuct the input `A' = U @ (S @ VT)`

File: crates/kornia-linalg/src/linalg.rs [hunk 4]
```diff
@@ -0,0 +1,488 @@
...
+    let z = -(*x);
+    if c {
+        *x = *y;
+        *y = z;
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+/// A simple symmetric 3x3 Matrix class (contains no storage for (0, 1) (0, 2) and (1, 2)
+pub struct Symmetric3x3 {
```
Comment (edgarriba) [inline crates/kornia-linalg/src/linalg.rs:71]:
```suggestion
struct Symmetric3x3 {
```

File: crates/kornia-linalg/src/linalg.rs [hunk 5]
```diff
@@ -0,0 +1,488 @@
...
+            m_20: mat.z_axis.x,
+            m_21: mat.z_axis.y,
+            m_22: mat.z_axis.z,
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+/// Helper struct to store 2 floats to avoid OUT parameters on functions
+pub struct Givens {
```
Comment (edgarriba) [inline crates/kornia-linalg/src/linalg.rs:119]:
```suggestion
struct Givens {
```

File: crates/kornia-linalg/src/linalg.rs [hunk 6]
```diff
@@ -0,0 +1,488 @@
...
+// Helper function to swap X and Y and swap Y with -X if c is true
+fn cond_neg_swap(c: bool, x: &mut f32, y: &mut f32) {
+    let z = -(*x);
+    if c {
+        *x = *y;
+        *y = z;
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
```
Comment (edgarriba) [inline crates/kornia-linalg/src/linalg.rs:69]: Double check that clone and copy traits are really needed. Same for the others

File: crates/kornia-linalg/src/linalg.rs [hunk 7]
```diff
@@ -0,0 +1,488 @@
...
+            3 => &mut self.0.w,
+            _ => panic!("Index out of bounds for Quaternion"),
+        }
+    }
+}
+
+/// Function used to apply a givens rotation S. Calculates the weights and updates the quaternion to contain the cumultative rotation
+fn jacobi_conjugation(x: usize, y: usize, z: usize, S: &mut Symmetric3x3, q: &mut IndexedQuat) {
+    // Compute the Givens rotation (approximated)
+    let mut g = approximate_givens_quaternion(S);
```
Comment (edgarriba) [inline crates/kornia-linalg/src/linalg.rs:240]:
```suggestion
    let mut g = approximate_givens_quaternion(S);

```

**Review by edgarriba — CHANGES_REQUESTED** [review]

File: crates/kornia-linalg/src/linalg.rs [hunk 8]
```diff
@@ -0,0 +1,488 @@
...
+            x_axis: Vec3::new(1.0, 0.0, 0.0),
+            y_axis: Vec3::new(0.0, 2.0, 0.0),
+            z_axis: Vec3::new(0.0, 0.0, 3.0),
+        };
+
+        // Perform SVD on matrix A
+        let A_clone = A.clone();
+        let svd_result = svd3(A_clone);
+        // Check matrix V
+        assert!(
```
[the remark this answers is not in the document]
Comment (prikshitsingh24) [reply crates/kornia-linalg/src/linalg.rs:471]: glam::Mat3 type doesn't implement the RelativeEq and AbsDiffEq

File: crates/kornia-linalg/src/linalg.rs [hunk 9]
```diff
@@ -0,0 +1,488 @@
...
+}
+
+#[cfg(test)]
+mod tests {
+    use glam::{Mat3, Vec3};
+
+    use super::*;
+
+    #[test]
+    fn test_svd3() {
```
[the remark this answers is not in the document]
Comment (prikshitsingh24) [reply crates/kornia-linalg/src/linalg.rs:459]: @edgarriba I wrote a bench file and these are the benchmarks:

svd3/svd3/ - time:   [1.3185 µs 1.3202 µs 1.3221 µs]
Found 6 outliers among 100 measurements (6.00%)
  1 (1.00%) low mild
  1 (1.00%) high mild
  4 (4.00%) high severe

svd3/svd3_faer/ - time:   [614.62 ns 618.62 ns 623.58 ns]
Found 19 outliers among 100 measurements (19.00%)
 7 (7.00%) high mild
 12 (12.00%) high severe

File: crates/kornia-linalg/src/linalg.rs [hunk 10]
```diff
@@ -0,0 +1,487 @@
...
+const RSQRT_STEPS: u8 = 4;
+const rsqrt_mut_STEPS: u8 = 6;
+
+/// Standard CPU division.
+fn fdiv(x: f32, y: f32) -> f32 {
+    return x / y;
+}
+
+/// Calculates the reciprocal square root of x using a fast approximation.
+fn rsqrt(x: f32) -> f32 {
```
Comment (edgarriba) [inline crates/kornia-linalg/src/linalg.rs:18]: We might want to try https://betterexplained.com/articles/understanding-quakes-fast-inverse-square-root/ which doesn’t have the loops. That could be the cause difference between faer ?

File: crates/kornia-linalg/src/linalg.rs [hunk 11]
```diff
@@ -0,0 +1,487 @@
...
+
+    Q.z_axis.x = 2.0 * g2.ch * g2.sh;
+    Q.z_axis.y = -2.0 * g3.ch * sh22 * g3.sh;
+    Q.z_axis.z = sh22 * sh32;
+
+    QR3 { Q, R }
+}
+
+/// Wrapping function used to contain all of the required sub calls
+pub fn svd3(A: Mat3) -> SVD3Set {
```
Comment (edgarriba) [inline crates/kornia-linalg/src/linalg.rs:428]:
```suggestion
pub fn svd3(A: &Mat3) -> SVD3Set {
```

File: crates/kornia-linalg/src/linalg.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (edgarriba) [reply crates/kornia-linalg/src/linalg.rs:458]: The only places where I could see the improvements are in the rsqrt, and in some places with micro optimisation to precompute some variables

File: crates/kornia-linalg/src/linalg.rs (hunk 10 above)
Comment (prikshitsingh24) [reply crates/kornia-linalg/src/linalg.rs:18]: Ok

Comment (prikshitsingh24) [reply crates/kornia-linalg/src/linalg.rs:18]: Current benchmark after changing the rsqrt:
Code_Og4rurVpvO

File: crates/kornia-linalg/src/linalg.rs [hunk 12]
```diff
@@ -0,0 +1,466 @@
...
+    S.m_20 = _S.m_20;
+    S.m_21 = _S.m_21;
+    S.m_22 = _S.m_22;
+}
+
+/// Function used to contain the givens permutations and the loop of the jacobi steps controlled by JACOBI_STEPS
+/// Returns the quaternion q containing the cumultative result used to reconstruct S
+fn jacobi_eigenanalysis(mut S: Symmetric3x3) -> Mat3 {
+    let mut q = IndexedQuat::new(Quat::from_xyzw(0.0, 0.0, 0.0, 1.0));
+    for _i in 0..JACOBI_STEPS {
```
Comment (edgarriba) [inline crates/kornia-linalg/src/linalg.rs:268]: Maybe play with the num steps to reduce and find balance between speed and results quality

File: crates/kornia-linalg/src/linalg.rs (hunk 10 above)
Comment (edgarriba) [reply crates/kornia-linalg/src/linalg.rs:18]: We are getting closer ! Let’s try to find out how to improve a little bit more

Comment (edgarriba) [reply crates/kornia-linalg/src/linalg.rs:18]: Also one suggestion while running benchmarks, try to have as less things as possible running in the background like web browser or slack, etc

Comment (prikshitsingh24) [reply crates/kornia-linalg/src/linalg.rs:18]: Ok got it

File: crates/kornia-linalg/src/linalg.rs (hunk 12 above)
Comment (prikshitsingh24) [reply crates/kornia-linalg/src/linalg.rs:268]: That worked changing JACOBI_STEPS from 12 to 6 is accurate and more speed than faer:
Code_itWQQcKALL

Comment (edgarriba) [reply crates/kornia-linalg/src/linalg.rs:271]: amazing! can we make sure with tests that the function produce proper results under some corner cases ? test with a couple different input matrices. Good job!

File: crates/kornia-linalg/src/linalg.rs [hunk 13]
```diff
@@ -0,0 +1,561 @@
+// Reference: https://github.com/wi-re/tbtSVD/blob/master/source/SVD.h
+use glam::{Mat3, Quat, Vec3};
```
Comment (edgarriba) [inline crates/kornia-linalg/src/linalg.rs:2]:
```suggestion
use glam::{Mat3, Quat};
```

File: crates/kornia-linalg/src/linalg.rs [hunk 14]
```diff
@@ -0,0 +1,561 @@
...
+
+    /// The sine of the angle in the Givens rotation.
+    sh: f32,
+}
+
+#[derive(Debug)]
+/// Helper struct to store 2 Matrices to avoid OUT parameters on functions
+struct QR3 {
+    /// The orthogonal matrix Q from the QR decomposition.
+    Q: Mat3,
```
Comment (edgarriba) [inline crates/kornia-linalg/src/linalg.rs:124]:
```suggestion
    q: Mat3,
```

Comment (edgarriba) [reply crates/kornia-linalg/src/linalg.rs:124]: let's use the lowecase notation. double check all the rest of the code

File: crates/kornia-linalg/src/linalg.rs [hunk 15]
```diff
@@ -0,0 +1,561 @@
...
+
+    /// The upper triangular matrix R from the QR decomposition.
+    R: Mat3,
```
Comment (edgarriba) [inline crates/kornia-linalg/src/linalg.rs:127]:
```suggestion
    r: Mat3,
```

File: crates/kornia-linalg/src/linalg.rs [hunk 16]
```diff
@@ -0,0 +1,561 @@
...
+}
+
+#[derive(Debug)]
+/// Helper struct to store 3 Matrices to avoid OUT parameters on functions
+pub struct SVD3Set {
+    /// The matrix of left singular vectors.
+    U: Mat3,
```
Comment (edgarriba) [inline crates/kornia-linalg/src/linalg.rs:134]:
```suggestion
    u: Mat3,
```

File: crates/kornia-linalg/src/linalg.rs [hunk 17]
```diff
@@ -0,0 +1,561 @@
...
+
+    /// The diagonal matrix of singular values.
+    S: Mat3,
```
Comment (edgarriba) [inline crates/kornia-linalg/src/linalg.rs:137]:
```suggestion
    s: Mat3,
```

File: crates/kornia-linalg/src/linalg.rs [hunk 18]
```diff
@@ -0,0 +1,561 @@
...
+
+    /// The matrix of right singular vectors.
+    V: Mat3,
```
Comment (edgarriba) [inline crates/kornia-linalg/src/linalg.rs:140]:
```suggestion
    v: Mat3,
```

File: crates/kornia-linalg/src/linalg.rs [hunk 19]
```diff
@@ -0,0 +1,561 @@
...
+
+/// Implementation of Algorithm 4
+fn qr_givens_quaternion(a1: f32, a2: f32) -> Givens {
+    // a1 = pivot point on diagonal
+    // a2 = lower triangular entry we want to annihilate
+    let epsilon = SVD_EPSILON; // Assuming _SVD_EPSILON is defined elsewhere
+    let rho = accurate_sqrt(a1 * a1 + a2 * a2);
+
+    let mut g = Givens {
+        ch: (a1.abs() + (f32::max(rho, epsilon))),
```
Comment (edgarriba) [inline crates/kornia-linalg/src/linalg.rs:321]:
```suggestion
        ch: a1.abs() + f32::max(rho, epsilon),
```

**Review by edgarriba — CHANGES_REQUESTED** [review]

**Review by edgarriba — COMMENTED** [review]
@prikshitsingh24 some test do not seems to success

**Comment by prikshitsingh24** [comment]
> @prikshitsingh24 some test do not seems to success

@edgarriba When i ported this in rust it gives this result for first test when I recompute the original matrix, is it how the algo should work like the answer from the algo should match the matrix A or there is some error in the code.
Since only the zero matrix test passes.
Left is the original A and right is the A computed by USV<sup>t</sup>
Code_fQnMxsJXTg

**Comment by edgarriba** [comment]
> > @prikshitsingh24 some test do not seems to success
> 
> @edgarriba When i ported this in rust it gives this result for first test when I recompute the original matrix, is it how the algo should work like the answer from the algo should match the matrix A or there is some error in the code. Since only the zero matrix test passes. Left is the original A and right is the A computed by USVt Code_fQnMxsJXTg

for sanity check, make sure that the improvements we made for example when computing the rsqrt (removing loops) didn't break the algorithm. Make also sure that glam matrices all well constructed and accessed. Maybe worth to add some other test from intermediate function to make sure everything makes sense.

**Comment by prikshitsingh24** [comment]
Hey @edgarriba, I found the issue. I reverted to my old code, and it worked fine. The problem was that the code I referred to used a matrix that stored data in row-major order, while I didn't realize that glam stores matrices in column-major order. Will Try to fix the glam part

**Comment by edgarriba** [comment]
@prikshitsingh24  rebase from main as I fixed the ci issues

**Review by edgarriba — APPROVED** [review]
